### PR TITLE
perf(sha256): Re-use hasher to reduce mallocs

### PR DESCRIPTION
**Pull request by itsdevbear**
## Summary by CodeRabbit

- **New Features**
  - Introduced new SHA256 hashing functionality for optimized performance.

- **Bug Fixes**
  - Updated various modules to use the new SHA256 package import for improved compatibility and performance.

- **Refactor**
  - Refactored test cases for readability and better structure.

- **Chores**
  - Updated dependencies and imports in multiple files to align with the latest package structure and versions.

### CODE REVIEW

File: mod/primitives/pkg/crypto/sha256/sha256.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	if !ok {
+		h = sha256.New()
+	}
+	defer sha256Pool.Put(h)
+	h.Reset()
+
+	var b [32]byte
+	//#nosec:G104 bet
+	h.Write(data)
+	h.Sum(b[:0])
```
Comment (rustybera) [inline mod/primitives/pkg/crypto/sha256/sha256.go:53]: > Sum appends the current hash to b and returns the resulting slice. It does not change the underlying hash state.

I see. `b[:0]` makes `Sum` think the slice is empty and write from the beginning.

File: mod/primitives/pkg/crypto/sha256/sha256.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+func CustomSHA256Hasher() func([]byte) [32]byte {
+	hasher, ok := sha256Pool.Get().(hash.Hash)
+	if !ok {
+		hasher = sha256.New()
+	} else {
+		hasher.Reset()
+	}
+	var h [32]byte
+
+	return func(data []byte) [32]byte {
```
Comment (rustybera) [inline mod/primitives/pkg/crypto/sha256/sha256.go:72]: Is this to reuse the memory slice?

Comment (itsdevbear) [reply mod/primitives/pkg/crypto/sha256/sha256.go:72]: yeah, in follow up PRs I think I am going to create hasher objs that are given to components of the system to re-use.

Comment (itsdevbear) [reply mod/primitives/pkg/crypto/sha256/sha256.go:72]: i.e state-processor should have it's own hasher.

Comment (rustybera) [reply mod/primitives/pkg/crypto/sha256/sha256.go:72]: This should prevent all race conditions?

**Review by rustybera — APPROVED** [review]
Lgtm.

**Pull request merged**